### PR TITLE
enhancement: returning an error when the popup is closed

### PR DESCRIPTION
**Pull request by dafuga**

### CODE REVIEW

File: src/account-creator.ts [hunk 1]
```diff
@@ -47,12 +48,29 @@ export class AccountCreator {
...
     }
 
+    cleanup() {
+        this.popupStatusInterval && clearInterval(this.popupStatusInterval)
+        this.popupStatusInterval = undefined
+        this.popupWindow = undefined
+    }
+
     closeDialog() {
         this.popupWindow?.close()
```
Comment (jnordberg) [inline src/account-creator.ts]: Cleanup not called here. Move the cleanup call here and use it as the canonical function for closing the window

File: Makefile [hunk 2]
```diff
@@ -30,7 +30,7 @@ ci-test: node_modules
 
 .PHONY: check
 check: node_modules
-	@${BIN}/eslint src --ext .ts --max-warnings 0 --format unix && echo "Ok"
+	@${BIN}/eslint src test --ext .ts --max-warnings 0 --format unix && echo "Ok"
```
Comment (jnordberg) [inline Makefile]: Why?

**Review by jnordberg — CHANGES_REQUESTED** [review]

File: Makefile (hunk 2 above)
Comment (dafuga) [reply Makefile]: Why not? I can remove it if you want, but I don't see why we wouldn't want to make sure that the tests also use proper ts types.

File: src/account-creator.ts (hunk 1 above)
Comment (dafuga) [reply src/account-creator.ts]: Done!

File: src/types.ts [hunk 3]
```diff
@@ -12,4 +12,4 @@ export interface AccountCreationErrorResponse {
     error: string
 }
 
-export type AccountCreationResponse = CallbackPayload | AccountCreationErrorResponse
+export type AccountCreationResponse = CallbackPayload | undefined
```
Comment (jnordberg) [inline src/types.ts]: This should be just CallbackPayload

File: Makefile (hunk 2 above)
Comment (jnordberg) [reply Makefile]: Type errors will be caught when running tests, this enforces additional strictness via eslint

File: src/account-creator.ts [hunk 4]
```diff
@@ -43,17 +44,38 @@ export class AccountCreator {
...
-        return new Promise((resolve) => {
+        return new Promise((resolve, reject) => {
             const listener = (event: MessageEvent) => {
                 window.removeEventListener('message', listener)
-                this.popupWindow?.close()
-                resolve(event.data)
+                this.closeDialog()
+
+                if (event.data.error) {
+                    reject(event.data)
```
Comment (jnordberg) [inline src/account-creator.ts]: This should be an Error object

**Review by jnordberg — CHANGES_REQUESTED** [review]

**Comment by dafuga** [comment]
@jnordberg Thanks! Made the changes!

File: README.md [hunk 5]
```diff
@@ -30,7 +30,9 @@ const accountCreator = new AccountCreator({
 })
 
 // Open a popup window prompting the user to create an account.
-const creationResult = await accountCreator.createAccount()
+const creationResult = await accountCreator.createAccount().catch((error) => {
```
Comment (jnordberg) [inline README.md]: `try { await promise } catch { .. }`
or
`promise.then(..).catch(..)`

never mix

**Review by jnordberg — CHANGES_REQUESTED** [review]

File: README.md (hunk 5 above)
Comment (dafuga) [reply README.md]: I'll make the change, but out of curiosity, what is wrong with  `await promise.catch(error => {})`. I've been doing that for ages 😄

Comment (dafuga) [reply README.md]: @jnordberg done!

Comment (jnordberg) [reply README.md]: Whatever is returned in the .catch block is returned as the creationResult so you end up having to handle the error twice by checking for undefined. TypeScript is good about forcing you to actually do that check but it's a common source of stray undefined vars floating around causing confusing errors.

Besides it's a bit offensive, like putting ketchup on icecream or something :P

Comment (jnordberg) [reply README.md]: The only valid use of `await promise.catch` I can think of is where you want to assign a default value if it errors:

```ts
let myVar = await promise.catch((error) => {
  console.warn('Unable to load my thing, using default', error)
  return defaultVar
})
```

But I'd argue its cleaner to write something like

```ts
let myVar
try {
  myVar = await promise
} catch (error) {
  console.warn('Unable to load my thing, using default', error)
  myVar = defaultVar
}
```

**Review by jnordberg — APPROVED** [review]

**Pull request merged**